src/Service/Service.php

```diff
@@ -0,0 +1,328 @@
+            /** @var ValidationError $vError */
+            $vError = $result->getFirstError();
+            throw ServiceException::invalidServiceData($vError);
+        }
+        return $result->isValid();
+    }
+
+    /**
+     * Delete all key/value pairs with empty value by recursively using array_filter
+     * @param array $input
+     * @return mixed[] array
+     */
+    private static function arrayFilterRec(array $input): array
+    {
+        foreach ($input as &$value) {
+            if (\is_array($value)) {
+                $value = self::arrayFilterRec($value);
+            }
+        }
+        return array_filter($input);
+    }
+
+    /**
+     * @return string
+     */
+    public function getServiceName(): string
+    {
+        return $this->serviceName;
+    }
+
+    /**
+     * @return string
+     */
+    public function getImage(): string
+    {
+        return $this->image;
+    }
+
+    /**
+     * @return int[]
+     */
+    public function getInternalPorts(): array
+    {
+        return $this->internalPorts;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getDependsOn(): array
+    {
+        return $this->dependsOn;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function getPorts(): array
+    {
+        return $this->ports;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function getLabels(): array
+    {
+        return $this->labels;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function getEnvironment(): array
+    {
+        return $this->environment;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    public function getVolumes(): array
+    {
+        return $this->volumes;
+    }
+
+    /**
+     * @param int $internalPort
+     */
+    public function addInternalPort(int $internalPort): void
+    {
+        $this->internalPorts[] = $internalPort;
+    }
+
+    /**
+     * @param string $dependsOn
+     */
+    public function addDependsOn(string $dependsOn): void
+    {
+        $this->dependsOn[] = $dependsOn;
+    }
+
+    /**
+     * @param mixed $port
+     */
+    public function addPort(array $port): void
+    {
+        $this->ports[] = array(
+            'source' => $port['source'] ?? '',  //optional
+            'target' => $port['target'],
+        );
+    }
+
+    /**
+     * @param mixed $label
+     */
+    public function addLabel(array $label): void
+    {
+        $this->labels[] = array(
+            'key' => $label['key'],
+            'values' => $label['values'] ?? '', //optional
+        );
+    }
+
+    /**
+     * @param mixed $environment
+     */
+    public function addEnvironment(array $environment): void
+    {
+        $this->environment[] = array(
+            'key' => $environment['key'],
+            'values' => $environment['values'] ?? '', //optional
+        );
+    }
+
+    /**
+     * @param mixed $volume
+     */
+    public function addVolume(array $volume): void
```

Comment:
Use addBindVolume, addTmpVolume, etc...

Reply:
Therefore, use "getBindVolumes", "getTmpVolumes"... We should split volumes in 3 parts.

Reply:
Create an object by Volume type